### PR TITLE
[WIP] implement "thermal compression" formulation

### PR DESCRIPTION
**Pull request by tjhei**
This is a new formulation for the mass conservation that adds the temperature dependency of the density gradient approximation.

Tests and benchmarks are currently missing.

### CODE REVIEW

**Comment by tjhei** [comment]
Oh, we should also discuss the name. Any suggestions are welcome.

File: source/postprocess/mass_error_statistics.cc [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+          statistics.add_value ("Integral mass error (kg/year)",
+                                global_integral_mass_error * year_in_seconds);
+          statistics.add_value ("Norm mass error (kg/year)",
+                                global_norm_mass_error * year_in_seconds);
+
+          // also make sure that the other columns filled by the this object
+          // all show up with sufficient accuracy and in scientific notation
+          {
+            const char *columns[] = { "Integral mass error (kg/year)",
+                                      "Norm mass error (kg/year)"
```
Comment (bangerth) [inline source/postprocess/mass_error_statistics.cc]: Norm OF mass error?

Same in the block below

File: source/postprocess/mass_error_statistics.cc [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                statistics.set_precision (columns[i], 8);
+                statistics.set_scientific (columns[i], true);
+              }
+          }
+        }
+      else
+        {
+          statistics.add_value ("Integral mass error (kg/s)",
+                                global_integral_mass_error);
+          statistics.add_value ("Norm mass error (kg/s)",
```
Comment (bangerth) [inline source/postprocess/mass_error_statistics.cc]: here

File: source/postprocess/mass_error_statistics.cc [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                                global_norm_mass_error);
+
+          // also make sure that the other columns filled by the this object
+          // all show up with sufficient accuracy and in scientific notation
+          {
+            const char *columns[] = { "Integral mass error (kg/s)",
+                                      "Norm mass error (kg/s)"
```
Comment (bangerth) [inline source/postprocess/mass_error_statistics.cc]: and here

File: source/postprocess/mass_error_statistics.cc [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        = Utilities::MPI::sum (local_integral_mass_error, this->get_mpi_communicator());
+      const double global_norm_mass_error
+        = std::sqrt(Utilities::MPI::sum (local_norm_mass_error, this->get_mpi_communicator()));
+
+      if (this->convert_output_to_years() == true)
+        {
+          statistics.add_value ("Integral mass error (kg/year)",
+                                global_integral_mass_error * year_in_seconds);
+          statistics.add_value ("Norm mass error (kg/year)",
+                                global_norm_mass_error * year_in_seconds);
```
Comment (bangerth) [inline source/postprocess/mass_error_statistics.cc]: see below

File: source/postprocess/mass_error_statistics.cc (hunk 1 above)
Comment (bangerth) [inline source/postprocess/mass_error_statistics.cc]: also, if you move these two lines one block up, then you don't need to write the same text twice

File: source/postprocess/mass_error_statistics.cc [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+              }
+          }
+        }
+
+      std::ostringstream output;
+      output.precision(3);
+      if (this->convert_output_to_years() == true)
+        output << global_integral_mass_error *year_in_seconds
+               << " kg/year, "
+               << global_norm_mass_error *year_in_seconds
```
Comment (bangerth) [inline source/postprocess/mass_error_statistics.cc]: add a space after `*` in lines 194 and 196

File: source/postprocess/mass_error_statistics.cc [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+namespace aspect
+{
+  namespace Postprocess
+  {
+    ASPECT_REGISTER_POSTPROCESSOR(MassErrorStatistics,
+                                  "mass error statistics",
+                                  "A postprocessor that computes some statistics about the "
+                                  "error of the mass conservation equation. More precisely, it "
+                                  "computes the norm and the integral of the equation "
+                                  "$\\rho \\Nabla \\mathbf u + \\Nabla \\rho \\cdot \\mathbf u$. "
```
Comment (bangerth) [inline source/postprocess/mass_error_statistics.cc]: `\nabla\cdot(\rho \mathbf u) = ...`

Lower-case nabla

File: source/postprocess/mass_error_statistics.cc [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+                                  "Therefore, its results indicate how accurate the chosen approximation "
+                                  "for compressibility is for the given model setup.")
```
Comment (bangerth) [inline source/postprocess/mass_error_statistics.cc]: Well, not quite. If you have inflow or outflow of the domain, then the result is *expected* to be nonzero!

File: source/postprocess/visualization/mass_error.cc [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                      density_gradient += out.densities[i] * fe_values_density.shape_grad(i,q);
+                      density += out.densities[i] * fe_values_density.shape_value(i,q);
+                    }
+
+                  rho_div_u  += density * velocity_divergence[q] * fe_values.JxW(q);
+                  u_grad_rho += velocity[q] * density_gradient * fe_values.JxW(q);
+                  cell_volume += fe_values.JxW(q);
+                }
+
+              (*return_value.second)(cell_index) = (rho_div_u + u_grad_rho) / cell_volume;
```
Comment (bangerth) [inline source/postprocess/visualization/mass_error.cc]: OK, so you compute *average* values on each cell. This should be part of the documentation.

File: source/postprocess/visualization/mass_error.cc [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    namespace VisualizationPostprocessors
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(MassError,
+                                                  "mass error",
+                                                  "A postprocessor that computes some statistics about the "
+                                                  "error of the mass conservation equation. More precisely, it "
+                                                  "computes the norm and the integral of the equation "
+                                                  "$\\rho \\Nabla \\mathbf u + \\Nabla \\rho \\cdot \\mathbf u$. "
+                                                  "Therefore, its results indicate how accurate the chosen approximation "
+                                                  "for compressibility is for the given model setup.")
```
Comment (bangerth) [inline source/postprocess/visualization/mass_error.cc]: This ought to be updated as discussed above

File: source/postprocess/visualization/mass_error_face_integral.cc [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        const unsigned int quadrature_degree = this->get_fe().base_element(this->introspection().base_elements.velocities).degree+1;
+
+        const QGauss<dim-1> quadrature_formula_face (quadrature_degree);
+        const QGauss<dim> quadrature_formula (quadrature_degree);
+
+        const unsigned int n_q_points = quadrature_formula.size();
+        const unsigned int n_q_points_face = quadrature_formula_face.size();
+
+        /* We need three different FEValues objects for the operations below.
+         * fe_values_support evaluates the material model at the support points
```
Comment (bangerth) [inline source/postprocess/visualization/mass_error_face_integral.cc]: this comment no longer seems accurate here

File: source/postprocess/visualization/mass_error_face_integral.cc [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  {
+    namespace VisualizationPostprocessors
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(MassErrorFaceIntegral,
+                                                  "mass error face integral",
+                                                  "A postprocessor that computes some statistics about the "
+                                                  "error of the mass conservation equation. More precisely, it "
+                                                  "computes the norm and the integral of the equation "
+                                                  "$\\rho \\Nabla \\mathbf u + \\Nabla \\rho \\cdot \\mathbf u$. "
+                                                  "Therefore, its results indicate how accurate the chosen approximation "
```
Comment (bangerth) [inline source/postprocess/visualization/mass_error_face_integral.cc]: same here regarding the comments I have above. should document the fact that you compute *average* values.

File: tests/thermal_compression/screen-output [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+      Relative Stokes residual after nonlinear iteration 13: 1.64026e-09
+   Rebuilding Stokes preconditioner...
+   Solving Stokes system... 18+0 iterations.
+      Relative Stokes residual after nonlinear iteration 14: 1.04143e-09
+   Rebuilding Stokes preconditioner...
+   Solving Stokes system... 14+0 iterations.
+      Relative Stokes residual after nonlinear iteration 15: 1.34868e-10
+   Rebuilding Stokes preconditioner...
+   Solving Stokes system... 12+0 iterations.
+      Relative Stokes residual after nonlinear iteration 16: 6.36475e-11
```
Comment (bangerth) [inline tests/thermal_compression/screen-output]: nice convergence of the nonlinear iteration :-)

**Comment by bangerth** [comment]
I don't particularly care for the name "thermal compression". It's a reaction to "isothermal compression" and describes how it differs from the latter, but it is not descriptive in what it actually does or what the underlying assumptions/approximations are.

**Comment by gassmoeller** [comment]
This was superseded by #2012.

**Pull request closed without merging**